zerver/migrations/0482_automatically_follow_unmute_topics_policy_defaults.py

```diff
@@ -0,0 +1,86 @@
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    RealmUserDefault = apps.get_model("zerver", "RealmUserDefault")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+
+    max_id_realm_user_default = RealmUserDefault.objects.aggregate(Max("id", default=0))["id__max"]
+    max_id_user_profile = UserProfile.objects.aggregate(Max("id", default=0))["id__max"]
+
+    BATCH_SIZE = 5000
+
+    lower_bound = 0
+    while lower_bound < max_id_realm_user_default:
+        RealmUserDefault.objects.filter(
+            id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE
+        ).update(
+            automatically_follow_topics_policy=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
+            automatically_unmute_topics_in_muted_streams_policy=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND,
+        )
+        lower_bound += BATCH_SIZE
+
+    lower_bound = 0
+    while lower_bound < max_id_user_profile:
+        UserProfile.objects.filter(id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE).update(
+            automatically_follow_topics_policy=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
+            automatically_unmute_topics_in_muted_streams_policy=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND,
+        )
+        lower_bound += BATCH_SIZE
+
+
+class Migration(migrations.Migration):
+    atomic = False
+    dependencies = [
+        ("zerver", "0481_alter_realm_uuid_alter_realm_uuid_owner_secret"),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name="realmuserdefault",
+            name="automatically_follow_topics_policy",
+            field=models.PositiveSmallIntegerField(
+                null=False,
+                default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
+            ),
+        ),
+        migrations.AlterField(
+            model_name="realmuserdefault",
+            name="automatically_unmute_topics_in_muted_streams_policy",
+            field=models.PositiveSmallIntegerField(
+                null=False,
+                default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND,
+            ),
+        ),
+        migrations.AlterField(
+            model_name="userprofile",
+            name="automatically_follow_topics_policy",
+            field=models.PositiveSmallIntegerField(
+                null=False,
+                default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
+            ),
+        ),
+        migrations.AlterField(
+            model_name="userprofile",
+            name="automatically_unmute_topics_in_muted_streams_policy",
+            field=models.PositiveSmallIntegerField(
+                null=False,
+                default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND,
+            ),
+        ),
+        migrations.RunPython(
+            set_default_user_topic_policies,
+            reverse_code=migrations.RunPython.noop,
+            elidable=True,
+        ),
```

Comment:
This migration is out of order. You have to call `set_default_user_topic_policies` before setting `null=False`, or those migration steps will fail.

Reply:
Changed the operations order

Reply:
Changed the operation order, but I think it'd not fail irrespective of the order.

When 0476 runs, and any new row is created before 0482. The value that would be stored in database will be what we have defined as default in models.py. NULL will not be stored. "Django doesn't set default at the database level, django ORM passes the default value as the value to store".

So throughout the migration process, any new row created will have the value stored as defined in models.py.

These migration changes are just cleaning up the history (how we changed from NEVER to new defaults and 0477 noop).

Let me know if I'm missing something?

I have verified what I've said above by deleting 0482 and creating a new user with default set to ON_SEND for both the fields.